applications/GeoMechanicsApplication/custom_elements/interface_element.cpp

```diff
@@ -339,65 +339,79 @@ void InterfaceElement::ApplyRotationToBMatrix(Matrix& rBMatrix, const Matrix& rR
-            rotation_matrix(2, 2) = 1.0;
-        } else {
-            rotation_matrix = mfpCalculateRotationMatrix(GetGeometry(), r_integration_point);
-        }
-
-        auto integration_point_stress_tensor = MathUtils<double>::StressVectorToTensor(integration_point_stress);
-        auto integration_point_local_stress_tensor = GeoMechanicsMathUtilities::RotateSecondOrderTensor(
-            integration_point_stress_tensor, rotation_matrix);
-
-        // extract normal and shear components to form initial traction
-        auto traction_vector = Vector(mConstitutiveLaws[0]->GetStrainSize());
-        if (r_interface_geometry.LocalSpaceDimension() == 1) {
-            traction_vector[0] = integration_point_local_stress_tensor(1, 1);
-            traction_vector[1] = integration_point_local_stress_tensor(0, 1);
-        } else {
-            traction_vector[0] = integration_point_local_stress_tensor(2, 2);
-            traction_vector[1] = integration_point_local_stress_tensor(0, 2);
-            traction_vector[2] = integration_point_local_stress_tensor(1, 2);
-        }
         const auto initial_state =
             make_intrusive<InitialState>(traction_vector, InitialState::InitialImposingType::STRESS_ONLY);
         mConstitutiveLaws[integration_point_index]->SetInitialState(initial_state);
         ++integration_point_index;
     }
 }
 
+Vector InterfaceElement::InterpolateNodalStressToIntegrationPoints(const Geo::IntegrationPointType& rIntegrationPoint,
+                                                                   std::vector<Vector> NodalStresses) const
+{
+    auto result                                  = Vector(NodalStresses[0].size(), 0.0);
+    auto integration_point_shape_function_values = Vector{};
+    GetGeometry().ShapeFunctionsValues(integration_point_shape_function_values, rIntegrationPoint);
+    for (auto i = std::size_t{0}; i < GetGeometry().PointsNumber() / 2; ++i) {
+        result += integration_point_shape_function_values[i] * NodalStresses[i];
+    }
+    return result;
+}
+
+Matrix InterfaceElement::RotateStressToLocalCoordinates(const Geo::IntegrationPointType& rIntegrationPoint,
+                                                        Vector& GlobalStressVector) const
+{
+    auto rotation_tensor = Matrix(3, 3, 0.0);
+    if (GetGeometry().LocalSpaceDimension() == 1) {
+        auto two_d_rotation_tensor = mfpCalculateRotationMatrix(GetGeometry(), rIntegrationPoint);
+        GeoElementUtilities::AddMatrixAtPosition(two_d_rotation_tensor, rotation_tensor, 0, 0);
+        rotation_tensor(2, 2) = 1.0;
+    } else {
+        rotation_tensor = mfpCalculateRotationMatrix(GetGeometry(), rIntegrationPoint);
+    }
+
+    auto integration_point_stress_tensor = MathUtils<>::StressVectorToTensor(GlobalStressVector);
+    auto result = GeoMechanicsMathUtilities::RotateSecondOrderTensor(integration_point_stress_tensor, rotation_tensor);
+    return result;
```

Comment:
Perhaps we can bring this back to a single statement:
```suggestion
    return GeoMechanicsMathUtilities::RotateSecondOrderTensor(MathUtils<>::StressVectorToTensor(GlobalStressVector), rotation_tensor);
```

Reply:
yes done